docs/ref/migration-operations.txt

```diff
@@ -402,7 +402,18 @@ You can take this template and work from it, though we suggest looking at the
 built-in Django operations in ``django.db.migrations.operations`` - they're
 easy to read and cover a lot of the example usage of semi-internal aspects
 of the migration framework like ``ProjectState`` and the patterns used to get
-historical models.
+historical models, as well as ``ModelState`` and the patterns used to mutate
+historical models in ``state_forwards``.
+
+.. warning::
```

Comment:
I don't think it's a common issue, but it can be extremely difficult to debug, and it's not obvious if you're not very familiar with this part of Django. Custom operations are not needed for 99% of the cases (even the example should be using `RunSQL`), so I think a good fat warning box for the people who do end up in this part of the documentation and need to use Django's internals is not a bad idea. 

I didn't put a lot of thought in the placement, I guess I could move it more towards the end, after the list of notes.